File: python/278_first_bad_version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
def isBadVersion(version):
    pass

class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        low = 1
        high = n
        
        while low < high:
            cur = (low + high) // 2
            if isBadVersion(cur):
                high = cur
            else:
                low = cur + 1

        return low

File: c++/278_first_bad_version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        long int l = 1, h = n;

        while (l <= h) {
            long int mid = (l + h) / 2;
            if (isBadVersion(mid)) {
                if (mid == 1 || !isBadVersion(mid - 1)) {
                    return mid;
                } else {
                    h = mid;
                }
            } else {
                l = mid + 1;
            }
        }

        return 1;
    }
};
